### ISSUE TITLE
[BUG] NodeJS out of memory error

### ISSUE THREAD

**Issue by trebor678** [microsoft/playwright-dotnet#2132]
**Context:**
- Playwright Version: 1.21.0
- Operating System: Windows
- .NET version: .NET 6
- Browser: Chrome v100.0.4896.127
- Extra: I am setting the location of Chrome to a local copy (v100.0.4896.127) rather than using Chromium shipped with Playwright

**The problem**
NodeJS is running out of memory

After leaving a few pages open via Playwright (each page sits in it's own context), NodeJSs memory is exceeded what it should and is then erroring out. The below image is after running for about 15 minutes.

![image](https://user-images.githubusercontent.com/26218881/167025366-f9049cd8-2017-4bed-ac7d-1651b71ac8e4.png)

If I leave it for much longer, I then get the following message

```
<--- Last few GCs --->

[8828:000001E4F537B5D0]  1604819 ms: Mark-sweep (reduce) 1883.3 (1891.3) -> 1883.0 (1890.8) MB, 16.2 / 0.0 ms  (+ 10.8 ms in 4 steps since start of marking, biggest step 10.7 ms, walltime since start of marking 31 ms) (average mu = 0.966, current mu = 0.9[8828:000001E4F537B5D0]  1604872 ms: Mark-sweep (reduce) 1886.2 (1893.0) -> 1883.8 (1891.7) MB, 31.0 / 0.0 ms  (+ 13.8 ms in 14 steps since start of marking, biggest step 13.2 ms, walltime since start of marking 48 ms) (average mu = 0.922, current mu = 0.

<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF671E3013F v8::internal::CodeObjectRegistry::~CodeObjectRegistry+112495
 2: 00007FF671DBF396 DSA_meth_get_flags+65526
 3: 00007FF671DC024D node::OnFatalError+301
 4: 00007FF6726F19EE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF6726DBECD v8::SharedArrayBuffer::Externalize+781
 6: 00007FF67257F61C v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF67257C754 v8::internal::Heap::CollectGarbage+4244
 8: 00007FF67257A0D0 v8::internal::Heap::AllocateExternalBackingStore+2000
 9: 00007FF672597D40 v8::internal::FreeListManyCached::Reset+1408
10: 00007FF6725983F5 v8::internal::Factory::AllocateRaw+37
11: 00007FF6725ADA2F v8::internal::FactoryBase<v8::internal::Factory>::NewRawTwoByteString+79
12: 00007FF67228AEAB v8::internal::StringStream::ClearMentionedObjectCache+9147
13: 00007FF6724AAB55 v8::sampler::Sampler::Stop+2629
14: 00007FF6724AC329 v8::sampler::Sampler::Stop+8729
15: 00007FF6724ACA29 v8::sampler::Sampler::Stop+10521
16: 00007FF6724AE420 v8::sampler::Sampler::Stop+17168
17: 00007FF6724B0D96 v8::base::TimeDelta::operator!=+7430
18: 00007FF67268B91F v8::internal::Builtins::code_handle+38911
19: 00007FF67277F4E1 v8::internal::SetupIsolateDelegate::SetupHeap+494673
20: 000001E4F6E580E6
```

**Comment by mxschmitt** [comment]
Can you share a repro with us?

Most likely you run into https://github.com/microsoft/playwright/issues/6319

Memory only gets flushed when the context gets closed.

**Comment by jvi-infare** [comment]
Hi all,

I believe that I had very similar issue which I've resolved by modifiying `lib\server\browserContext.js` and removing context listener from instrumentation map.

I've noticed that when new a context is created, it's being inserted into instrumentation map (`lib\server\instrumentation.js`) by calling `this.instrumentation.addListener(contextDebugger, this);` but when it's closed, removeListener is never called (or at least that's happening it my case) so context stays in the map and memory won't get released.

I made some testing for ~12 hours by creating, utilizing and then closing thousands of contexts and they all left hanging in this map even though they were all closed and only "parent" browser was left opened. All the browser tabs were also closed, except for the "parent" one. NodeJS was consuming gygabytes of RAM and once I removed all the listeners, memory was flushed.

I can try to provide more details if you think that this might be the case.
